utils/interned_string.h

```diff
@@ -0,0 +1,222 @@
+/*****************************************************************************/
+
+template<size_t Bytes, typename Char = char>
+struct InternedString {
+
+    // Max Bytes is 255, as otherwise we can't determine the length
+    static_assert(Bytes < 256, "First template parameter for InternedString "
+                  "must be 255 or less");
+
+    InternedString()
+        : intLength_(0)
+    {
+    }
+
+    InternedString(const InternedString & other)
+        : InternedString()
+    {
+        append(other.data(), other.length());
+    }
+
+    template<size_t OtherBytes>
+    InternedString(const InternedString<OtherBytes, Char> & other)
+        : InternedString()
+    {
+        append(other.data(), other.length());
+    }
+
+    template<size_t OtherBytes>
+    InternedString(InternedString<OtherBytes, Char> && other) noexcept
+        : InternedString()
+    {
+        if (other.length() > Bytes) {
+            // Can't fit internally.  If the other is external, steal it
+            if (other.isExt()) {
+                intLength_ = IS_EXT;
+                extLength_ = other.extLength_;
+                extCapacity_ = other.extCapacity_;
+                extBytes_ = other.extBytes_;
+                other.intLength_ = 0;
+                return;
+            }
+        }
+
+        // Otherwise, simply append it
+        append(other.data(), other.size());
+    }
+
+    InternedString(const std::basic_string<Char> & other)
+        : InternedString()
+    {
+        append(other.data(), other.length());
+    }
+
+    InternedString & operator = (const InternedString & other)
+    {
+        InternedString newMe(other);
+        swap(newMe);
+        return *this;
+    }
+
+    InternedString & operator = (InternedString && other) noexcept
+    {
+        InternedString newMe(std::move(other));
+        swap(newMe);
+        return *this;
+    }
+
+    ~InternedString()
+    {
+        if (isExt())
+            deleteExt();
+    }
+
+    const Char * data() const
+    {
+        return isExt() ? extBytes_ : intBytes_;
+    }
+
+    void swap(InternedString & other) noexcept
+    {
+        std::swap(intLength_, other.intLength_);
+        std::swap(intBytes_[0], other.intBytes_[0]);
+        std::swap(intBytes_[1], other.intBytes_[1]);
+        std::swap(intBytes_[2], other.intBytes_[2]);
+        for (unsigned i = 0;  i < NUM_WORDS * 2 - 1;  ++i) {
+            std::swap(internalWords[i], other.internalWords[i]);
+        }
+    }
+
+    size_t size() const
+    {
+        return isExt() ? extLength_ : intLength_;
+    }
+
+    size_t length() const
+    {
+        return size();
+    }
+
+    bool empty() const
+    {
+        return size() == 0;
+    }
+
+    void reserve(size_t newCapacity)
+    {
+        if (newCapacity < capacity())
+            return;
+
+        bool wasExt = isExt();
+
+        char * newBytes = new Char[newCapacity];
+        size_t l = size();
```

Comment:
Or if this is a "no except" section, size() should also be declared with "noexcept".
If it is the case, I would surroumd the section with clear comments that this is a "memory safe zone".

Reply:
You are right, but again this moves outside of the scope (it was a code move).

Reply:
I have added a comment and a static_assert for the PODness to make it clear.